### PR TITLE
Beacd-54: Under - Subcontractor journey

### PR DESCRIPTION
**Pull request by prashantchoudhari**

### CODE REVIEW

File: app/controllers/employer/cis/uk/subcontractor/DoYouWantToBePaidNetOrGrossController.scala [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    with I18nSupport
+    with Enumerable.Implicits {
+
+  val form = formProvider()
+
+  def onPageLoad() = (authenticate andThen serviceInfoData) { implicit request =>
+    Ok(doYouWantToBePaidNetOrGross(appConfig, form)(request.serviceInfoContent))
+  }
+
+  def onSubmit() = (authenticate andThen serviceInfoData).async { implicit request =>
```
Comment (kurtlogan) [inline app/controllers/employer/cis/uk/subcontractor/DoYouWantToBePaidNetOrGrossController.scala]: .async can be removed

File: app/controllers/employer/cis/uk/subcontractor/WasTurnoverMoreAfterVATController.scala [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    with I18nSupport
+    with Enumerable.Implicits {
+
+  val form = formProvider()
+
+  def onPageLoad() = (authenticate andThen serviceInfoData) { implicit request =>
+    Ok(wasTurnoverMoreAfterVAT(appConfig, form)(request.serviceInfoContent))
+  }
+
+  def onSubmit() = (authenticate andThen serviceInfoData).async { implicit request =>
```
Comment (kurtlogan) [inline app/controllers/employer/cis/uk/subcontractor/WasTurnoverMoreAfterVATController.scala]: .async can be removed

File: app/controllers/employer/cis/uk/subcontractor/WhatTypeOfSubcontractorController.scala [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    with I18nSupport
+    with Enumerable.Implicits {
+
+  val form = formProvider()
+
+  def onPageLoad() = (authenticate andThen serviceInfoData) { implicit request =>
+    Ok(whatTypeOfSubcontractor(appConfig, form)(request.serviceInfoContent))
+  }
+
+  def onSubmit() = (authenticate andThen serviceInfoData).async { implicit request =>
```
Comment (kurtlogan) [inline app/controllers/employer/cis/uk/subcontractor/WhatTypeOfSubcontractorController.scala]: .async can be removed

File: app/utils/nextpage/employer/cis/uk/subcontractor/WhatTypeOfSubcontractorNextPage.scala [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+  type WhatTypeOfSubcontractorWithEnrolments = (WhatTypeOfSubcontractor, Enrolments)
+
+  implicit val whatTypeOfSubcontractor
+    : NextPage[WhatTypeOfSubcontractorId.type, WhatTypeOfSubcontractorWithEnrolments] = {
+    new NextPage[WhatTypeOfSubcontractorId.type, WhatTypeOfSubcontractorWithEnrolments] {
+      override def get(
+        b: WhatTypeOfSubcontractorWithEnrolments)(implicit appConfig: FrontendAppConfig, request: Request[_]): Call = {
+
+        val (fst, second) = b
```
Comment (kurtlogan) [inline app/utils/nextpage/employer/cis/uk/subcontractor/WhatTypeOfSubcontractorNextPage.scala]: Rename variables to something more meaningful

File: conf/messages.cy [hunk 5]
```diff
@@ -642,6 +642,37 @@ registerInHomeCountry.p1=Dylech gofrestru ar gyfer GUC TAW gyda’r awdurdod tre
 registerInHomeCountry.p2=Os ydych eisiau cofrestru ar gyfer GUC TAW yn y DU, mae angen i chi
 registerInHomeCountry.callUs = ein ffonio.
 
+#######################################################
+##  DoYouWantToBePaidNetOrGross
+#######################################################
+doYouWantToBePaidNetOrGross.title = A ydych eisiau cael eich talu’n net neu’n gros?
+doYouWantToBePaidNetOrGross.heading = A ydych eisiau cael eich talu’n net neu’n gros?
+doYouWantToBePaidNetOrGross.subtext1 = Os cewch eich talu’n net, caiff 20% o’ch cyflog ei ddidynnu at ddibenion treth ac Yswiriant Gwladol.
+doYouWantToBePaidNetOrGross.subtext2 = Os ydych am gael eich talu’n gros, rhaid i chi {0} cyn mynd ymlaen
```
Comment (kurtlogan) [inline conf/messages.cy]: Was it confirmed that this was the right place to put the anchor tag?

File: test/utils/nextpage/employer/cis/uk/subcontractor/WhatTypeOfSubcontractorNextPageSpec.scala [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+    behave like nextPageWithEnrolments(
+      NextPage.whatTypeOfSubcontractor,
+      nextPageLimitedCompanyNoValue,
+      WhatTypeOfSubcontractor.LimitedCompany.toString,
+      "https://www.gov.uk/government/publications/construction-industry-scheme-company-registration-cis305",
+      "no Enrolment"
+    )
+
+    /*
```
Comment (kurtlogan) [inline test/utils/nextpage/employer/cis/uk/subcontractor/WhatTypeOfSubcontractorNextPageSpec.scala]: Why has this been commented out. If it is not needed then remove it. If it is needed than uncomment it.

File: test/views/employer/cis/uk/subcontractor/DoYouWantToBePaidNetOrGrossViewSpec.scala [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        "If you’re paid net, 20% of your pay will be deducted to cover tax and National Insurance.")
+    }
+
+    "contain the subtext2" in {
+      val doc = asDocument(createView()).text()
+
+      doc must include(
+        "If you want to be paid gross, you must check if you qualify for gross payments before you continue.")
+    }
+    "contains check if you qualify for gross payment - link" in {
```
Comment (kurtlogan) [inline test/views/employer/cis/uk/subcontractor/DoYouWantToBePaidNetOrGrossViewSpec.scala]: add a space between these two blocks

File: test/views/employer/cis/uk/subcontractor/WasTurnoverMoreAfterVATViewSpec.scala [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+      }
+    }
+
+    for (option <- WasTurnoverMoreAfterVAT.options) {
+      s"rendered with a value of '${option.value}'" must {
+        s"have the '${option.value}' radio button selected" in {
+          val doc = asDocument(createViewUsingForm(form.bind(Map("value" -> s"${option.value}"))))
+          assertContainsRadioButton(doc, option.id, "value", option.value, true)
+
+          for (unselectedOption <- WasTurnoverMoreAfterVAT.options.filterNot(o => o == option)) {
```
Comment (kurtlogan) [inline test/views/employer/cis/uk/subcontractor/WasTurnoverMoreAfterVATViewSpec.scala]: You can change the lamba to be `_ == option` it is a little more expressive and there is less noise

File: test/views/employer/cis/uk/subcontractor/WhatTypeOfSubcontractorViewSpec.scala [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+      }
+    }
+
+    for (option <- WhatTypeOfSubcontractor.options) {
+      s"rendered with a value of '${option.value}'" must {
+        s"have the '${option.value}' radio button selected" in {
+          val doc = asDocument(createViewUsingForm(form.bind(Map("value" -> s"${option.value}"))))
+          assertContainsRadioButton(doc, option.id, "value", option.value, true)
+
+          for (unselectedOption <- WhatTypeOfSubcontractor.options.filterNot(o => o == option)) {
```
Comment (kurtlogan) [inline test/views/employer/cis/uk/subcontractor/WhatTypeOfSubcontractorViewSpec.scala]: As above, the lamda can be replaced with `_ == option`

File: conf/messages.cy (hunk 5 above)
Comment (prashantchoudhari) [reply conf/messages.cy]: Yes, as per ticket

File: app/controllers/employer/cis/uk/subcontractor/DoYouWantToBePaidNetOrGrossController.scala (hunk 1 above)
Comment (prashantchoudhari) [reply app/controllers/employer/cis/uk/subcontractor/DoYouWantToBePaidNetOrGrossController.scala]: Lets see if we can modify template!!

File: app/controllers/employer/cis/uk/subcontractor/WasTurnoverMoreAfterVATController.scala (hunk 2 above)
Comment (prashantchoudhari) [reply app/controllers/employer/cis/uk/subcontractor/WasTurnoverMoreAfterVATController.scala]: Lets see if we can modify template!!

File: app/controllers/employer/cis/uk/subcontractor/WhatTypeOfSubcontractorController.scala (hunk 3 above)
Comment (prashantchoudhari) [reply app/controllers/employer/cis/uk/subcontractor/WhatTypeOfSubcontractorController.scala]: Lets see if we can modify template!!

File: app/utils/nextpage/employer/cis/uk/subcontractor/WhatTypeOfSubcontractorNextPage.scala (hunk 4 above)
Comment (prashantchoudhari) [reply app/utils/nextpage/employer/cis/uk/subcontractor/WhatTypeOfSubcontractorNextPage.scala]: good catch..

File: test/utils/nextpage/employer/cis/uk/subcontractor/WhatTypeOfSubcontractorNextPageSpec.scala (hunk 6 above)
Comment (prashantchoudhari) [reply test/utils/nextpage/employer/cis/uk/subcontractor/WhatTypeOfSubcontractorNextPageSpec.scala]: 👍

File: test/views/employer/cis/uk/subcontractor/WasTurnoverMoreAfterVATViewSpec.scala (hunk 8 above)
Comment (prashantchoudhari) [reply test/views/employer/cis/uk/subcontractor/WasTurnoverMoreAfterVATViewSpec.scala]: Lets see if we can modify template!!

File: test/views/employer/cis/uk/subcontractor/WhatTypeOfSubcontractorViewSpec.scala (hunk 9 above)
Comment (prashantchoudhari) [reply test/views/employer/cis/uk/subcontractor/WhatTypeOfSubcontractorViewSpec.scala]: Lets see if we can modify template!!

File: app/utils/nextpage/employer/cis/uk/subcontractor/WhatTypeOfSubcontractorNextPage.scala [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+  implicit val whatTypeOfSubcontractor
+    : NextPage[WhatTypeOfSubcontractorId.type, WhatTypeOfSubcontractorWithEnrolments] = {
+    new NextPage[WhatTypeOfSubcontractorId.type, WhatTypeOfSubcontractorWithEnrolments] {
+      override def get(
+        b: WhatTypeOfSubcontractorWithEnrolments)(implicit appConfig: FrontendAppConfig, request: Request[_]): Call = {
+
+        val (whatTypeOfSubcontractor, enrolments) = b
+        val hasEnrolments: Boolean =
+          utils.Enrolments.hasEnrolments(enrolments, Seq(HmrcEnrolmentType.SA, HmrcEnrolmentType.CORP_TAX): _*)
```
Comment (kurtlogan) [inline app/utils/nextpage/employer/cis/uk/subcontractor/WhatTypeOfSubcontractorNextPage.scala]: `hasEnrolments` is a variadic function so you don't need to wrap the enrolment types in a seq, you can provide them as individual arguments like so 
```
utils.Enrolments.hasEnrolments(enrolments, HmrcEnrolmentType.SA, HmrcEnrolmentType.CORP_TAX)
```

File: conf/application.conf [hunk 11]
```diff
@@ -186,6 +187,12 @@ urls {
     warehousing = "/government/publications/excise-warehousing-application-to-be-an-excise-warehousekeeper-ex61"
     charity = "/charity-recognition-hmrc"
     vatEnquiries = "/government/organisations/hm-revenue-customs/contact/vat-enquiries"
+    cisRegisterPaidNet = "/government/publications/construction-industry-scheme-individual-registration-for-payment-under-deduction-cis301"
+    cisRegisterPaidGross = "/government/publications/construction-industry-scheme-individual-registration-for-gross-payment-cis302"
+    cisRegisterLowTurnover = "/government/publications/construction-industry-scheme-individual-registration-for-gross-payment-cis302"
```
Comment (kurtlogan) [inline conf/application.conf]: This config is the same as the one above, you should reused the same config

File: test/views/employer/cis/uk/subcontractor/DoYouWantToBePaidNetOrGrossViewSpec.scala [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+      }
+    }
+
+    for (option <- DoYouWantToBePaidNetOrGross.options) {
+      s"rendered with a value of '${option.value}'" must {
+        s"have the '${option.value}' radio button selected" in {
+          val doc = asDocument(createViewUsingForm(form.bind(Map("value" -> s"${option.value}"))))
+          assertContainsRadioButton(doc, option.id, "value", option.value, true)
+
+          for (unselectedOption <- DoYouWantToBePaidNetOrGross.options.filterNot(o => o == option)) {
```
Comment (kurtlogan) [inline test/views/employer/cis/uk/subcontractor/DoYouWantToBePaidNetOrGrossViewSpec.scala]: Replace lambda with `_ == option`

**Pull request merged**